core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -720,15 +728,24 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
     return builder.build();
   }
 
-  private Map<String, SnapshotRef> validateRefs(Map<String, SnapshotRef> inputRefs) {
-    for (Map.Entry<String, SnapshotRef> refEntry : inputRefs.entrySet()) {
-      String refName = refEntry.getKey();
-      SnapshotRef ref = refEntry.getValue();
-      if (!(ref.snapshotId() == -1 && refName.equals(SnapshotRef.MAIN_BRANCH))) {
-        Preconditions.checkArgument(snapshotsById.containsKey(ref.snapshotId()),
-            "Snapshot reference %s does not exist in the existing snapshots list", ref);
-      }
+  private static Map<String, SnapshotRef> validateRefs(Long currentSnapshotId,
+                                                       Map<String, SnapshotRef> inputRefs,
+                                                       Map<Long, Snapshot> snapshotsById) {
+    for (SnapshotRef ref : inputRefs.values()) {
+      Preconditions.checkArgument(snapshotsById.containsKey(ref.snapshotId()),
+          "Snapshot for reference %s does not exist in the existing snapshots list", ref);
     }
+
+    SnapshotRef main = inputRefs.get(SnapshotRef.MAIN_BRANCH);
+    if (currentSnapshotId != -1) {
+      Preconditions.checkArgument(main == null || currentSnapshotId == main.snapshotId(),
+          "Current snapshot ID does not match main branch (%s != %s)",
+          currentSnapshotId, main != null ? main.snapshotId() : null);
+    } else {
+      Preconditions.checkArgument(main == null,
+          "Current snapshot is not set, but main branch exists: %s", main);
```

Comment:
nit: should this mention `...snapshot ID is not set...` to be consistent with the error message above?